src/firebase_functions/db.py

```diff
@@ -0,0 +1,211 @@
+import cloudevents.http as _ce
+from firebase_functions.core import Change  # Exported for user ease of typing events
+
+_event_type_written = "google.firebase.database.ref.v1.written"
+_event_type_created = "google.firebase.database.ref.v1.created"
+_event_type_updated = "google.firebase.database.ref.v1.updated"
+_event_type_deleted = "google.firebase.database.ref.v1.deleted"
+
+
+@_dataclass.dataclass(frozen=True)
+class DatabaseEvent(_core.CloudEvent[_core.T]):
+    """
+    A CloudEvent that contains a DataSnapshot or a Change<DataSnapshot>.
+    """
+
+    firebase_database_host: str
+    """
+    The domain of the database instance.
+    """
+
+    instance: str
+    """
+    The instance ID portion of the fully qualified resource name.
+    """
+
+    reference: str
+    """
+    The database reference path.
+    """
+
+    location: str
+    """
+    The location of the database
+    """
+
+    params: dict[str, str]
+    """
+    An dict containing the values of the path patterns.
+    Only named capture groups are populated - {key}, {key=*}, {key=**}
+    """
+
+
+_E1 = DatabaseEvent[Change[_typing.Any | None]]
+_E2 = DatabaseEvent[_typing.Any | None]
+_C1 = _typing.Callable[[_E1], None]
+_C2 = _typing.Callable[[_E2], None]
+
+
+def _db_endpoint_handler(
+    func: _C1 | _C2,
+    raw: _ce.CloudEvent,
+) -> None:
+    event_attributes = raw._get_attributes()
+    event_data = raw._get_data()
+    # TODO Params are built locally via path pattern which is currently unimplemented
+    params: dict[str, str] = {}
+    database_event_data = event_data["data"]
+    if "delta" in event_data:
+        before = event_data["data"]
+        after = event_data["delta"]
+        # Merge delta into data to generate an 'after' view of the data.
+        if isinstance(before, dict) and isinstance(after, dict):
+            after = _util.prune_nones({**before, **after})
+        database_event_data = _core.Change(
+            before=before,
+            after=after,
+        )
+    database_event = DatabaseEvent(
+        firebase_database_host=event_attributes["firebasedatabasehost"],
+        instance=event_attributes["instance"],
+        reference=event_attributes["ref"],
+        location=event_attributes["location"],
+        specversion=event_attributes["specversion"],
+        id=event_attributes["id"],
+        source=event_attributes["source"],
+        type=event_attributes["type"],
+        time=_dt.datetime.strptime(
+            event_attributes["time"],
+            "%Y-%m-%dT%H:%M:%S.%f%z",
+        ),
+        data=database_event_data,
+        subject=event_attributes["subject"],
+        params=params,
+    )
+    func(database_event)
+
+
+@_util.copy_func_kwargs(_options.DatabaseOptions)
+def on_value_written(**kwargs) -> _typing.Callable[[_C1], _C1]:
+    """
+    Event handler which triggers when data is created, updated, or deleted in Realtime Database.
+    Example::
+      @on_value_written(reference="*")
```

Comment:
nit: "*" makes a bad reference example. If someone runs it and there's a large database they could impact production and it doesn't show off parameters which are kinda the whole point. Maybe something like `"messages/{id}"` or `"users/{uid}"`